src/txmempool.cpp

```diff
@@ -649,10 +639,7 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
     }
     RemoveStaged(setAllRemoves, false, MemPoolRemovalReason::REORG);
     for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-        const LockPoints lp{it->GetLockPoints()};
-        if (!TestLockPointValidity(chain, lp)) {
-            mapTx.modify(it, update_lock_points(lp));
-        }
+        assert(TestLockPointValidity(chain, it->GetLockPoints()));
```

Comment:
Is it not better to add this assert as soon as the data is updated? That is, something like this:

<details>
<summary>patch</summary>

```diff
diff --git i/src/txmempool.cpp w/src/txmempool.cpp
index 26cb2ff4b4..8149c6dcd7 100644
--- i/src/txmempool.cpp
+++ w/src/txmempool.cpp
@@ -635,15 +635,12 @@ void CTxMemPool::removeForReorg(CChain& chain, std::function<bool(txiter)> check
     }
     setEntries setAllRemoves;
     for (txiter it : txToRemove) {
         CalculateDescendants(it, setAllRemoves);
     }
     RemoveStaged(setAllRemoves, false, MemPoolRemovalReason::REORG);
-    for (indexed_transaction_set::const_iterator it = mapTx.begin(); it != mapTx.end(); it++) {
-        assert(TestLockPointValidity(chain, it->GetLockPoints()));
-    }
 }
 
 void CTxMemPool::removeConflicts(const CTransaction &tx)
 {
     // Remove transactions which depend on inputs of tx, recursively
     AssertLockHeld(cs);
diff --git i/src/validation.cpp w/src/validation.cpp
index 68729e4863..541b7be499 100644
--- i/src/validation.cpp
+++ w/src/validation.cpp
@@ -377,12 +377,15 @@ void CChainState::MaybeUpdateMempoolForReorg(
                     break;
                 }
             }
         }
         // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
         if (!validLP) m_mempool->mapTx.modify(it, update_lock_points(lp));
+        if (!should_remove) {
+            assert(TestLockPointValidity(m_chain, it->GetLockPoints()));
+        }
         return should_remove;
     };
 
     // We also need to remove any now-immature transactions
     m_mempool->removeForReorg(m_chain, check_final_and_mature);
     // Re-limit mempool size, in case we added any transactions
```
</details>

Reply:
Seems better to run it on all remaining txs unconditionally before returning as a sanity check.

Reply:
I agree it seems like a better sanity check run at the end of `removeForReorg`, once all of the entries have been processed